src/relay/backend/build_module.cc

```diff
@@ -0,0 +1,669 @@
+  static const std::unordered_map<std::string, int> str2mask;
+  static std::string Mask2Str(int mask) {
+    CHECK_GT(mask2str.count(mask), 0) << "Unknown mask.";
+    return mask2str.at(mask);
+  }
+  static int Str2Mask(const std::string& str) {
+    CHECK_GT(str2mask.count(str), 0) << "Unknown context.";
+    return str2mask.at(str);
+  }
+  static std::unordered_map<int, std::string> _declare_mask2str() {
+    std::unordered_map<int, std::string> ret;
+    ret[1] = "cpu";
+    ret[2] = "gpu";
+    ret[4] = "opencl";
+    ret[5] = "aocl";
+    ret[6] = "sdaccel";
+    ret[7] = "valkan";
+    ret[8] = "metal";
+    ret[9] = "vpi";
+    ret[10] = "rocm";
+    ret[11] = "opengl";
+    ret[12] = "ext_dev";
+    return ret;
+  }
+  static std::unordered_map<std::string, int> _declare_str2mask() {
+    std::unordered_map<std::string, int> ret;
+    ret["llvm"] = 1;
+    ret["stackvm"] = 1;
+    ret["cpu"] = 1;
+    ret["c"] = 1;
+    ret["gpu"] = 2;
+    ret["cuda"] = 2;
+    ret["nvptx"] = 2;
+    ret["cl"] = 4;
+    ret["opencl"] = 4;
+    ret["aocl"] = 5;
+    ret["aocl_sw_emu"] = 5;
+    ret["sdaccel"] = 6;
+    ret["vulkan"] = 7;
+    ret["metal"] = 8;
+    ret["vpi"] = 9;
+    ret["rocm"] = 10;
+    ret["opengl"] = 11;
+    ret["ext_dev"] = 12;
+    return ret;
+  }
+};
+
+const std::unordered_map<int, std::string> ContextMap::mask2str =
+  ContextMap::_declare_mask2str();
+const std::unordered_map<std::string, int> ContextMap::str2mask =
+  ContextMap::_declare_str2mask();
+
+/*! \brief Optimization pass level */
+struct OptPassLevel {
+  static const std::unordered_map<std::string, int> _data;
+  static std::unordered_map<std::string, int> _declare_opt_level() {
+    std::unordered_map<std::string, int> ret;
+    ret["SimplifyInference"] = 0;
+    ret["OpFusion"] = 1;
+    ret["FoldConstant"] = 2;
+    ret["CombineParallelConv2D"] = 3;
+    ret["FoldScaleAxis"] = 3;
+    ret["AlterOpLayout"] = 3;
+    ret["CanonicalizeOps"] = 3;
+    ret["EliminateCommonSubexpr"] = 3;
+    return ret;
+  }
+  int operator[](const std::string& key) const {
+    auto it = _data.find(key);
+    if (it == _data.end()) {
+      return -1;
+    }
+    return it->second;
+  }
+};
+
+const std::unordered_map<std::string, int> OptPassLevel::_data =
+  OptPassLevel::_declare_opt_level();
+
+/*! \brief Output of function building */
+struct BuildOutput {
+  std::string graph_json;
+  runtime::Module mod;
+  std::unordered_map<std::string, tvm::runtime::NDArray> params;
+};
+
+/*! \brief Relay Building configuration */
+struct RelayBuildConfig {
+  int opt_level{2};
+  std::string fall_back_device{"llvm"};
+  std::unordered_set<std::string> add_pass;
+  std::unordered_set<std::string> disabled_pass;
+  OptPassLevel  OPT_PASS_LEVEL;
+  inline bool pass_enabled(std::string pass_name) const {
+    if (disabled_pass.count(pass_name)) {
+      return false;
+    }
+    if (add_pass.count(pass_name)) {
+      return true;
+    }
```

Comment:
check that if something is in disabled_pass, it is not in add_pass

Reply:
This logic is correct.